File: ocaml/lib/session.ml
(* Copyright (c) Microsoft Corporation. All rights reserved. *)

(** Copilot session -- manages a single conversation. *)

(* ========================================================================== *)
(* Types                                                                      *)
(* ========================================================================== *)

type event_handler = Types.session_event -> unit Lwt.t

type tool_handler =
  Types.tool_invocation -> Types.tool_result Lwt.t

type permission_handler =
  Yojson.Safe.t -> string -> Types.permission_result Lwt.t

type exit_plan_mode_handler =
  Types.exit_plan_mode_request -> Types.exit_plan_mode_response Lwt.t

type t = {
  rpc : Jsonrpc.t;
  session_id : string;
  config : Types.session_config;
  mutable event_handlers : event_handler list;
  tool_handlers : (string, tool_handler) Hashtbl.t;
  mutable permission_handler : permission_handler option;
  mutable exit_plan_mode_handler : exit_plan_mode_handler option;
  mutable trace_context_provider : (unit -> Types.trace_context) option;
  mutable idle : bool;
}

(* ========================================================================== *)
(* Construction                                                               *)
(* ========================================================================== *)

let create ~rpc ~session_id ~config =
  let t =
    { rpc
    ; session_id
    ; config
    ; event_handlers = []
    ; tool_handlers = Hashtbl.create 8
    ; permission_handler = None
    ; exit_plan_mode_handler = None
    ; trace_context_provider = None
    ; idle = false
    }
  in
  (* Register JSON-RPC handlers for tool calls and permissions *)
  Jsonrpc.register_handler rpc "tools/handlePendingToolCall"
    (fun _method params ->
      let open Lwt.Syntax in
      match Types.tool_invocation_of_yojson params with
      | Error msg ->
        Lwt.return
          (Types.tool_result_to_yojson
             (Types.make_tool_result ~result_type:Failure msg))
      | Ok invocation ->
        (match Hashtbl.find_opt t.tool_handlers invocation.tool_name with
         | None ->
           Lwt.return
             (Types.tool_result_to_yojson
                (Types.make_tool_result ~result_type:Failure
                   (Printf.sprintf "No handler for tool: %s"
                      invocation.tool_name)))
         | Some handler ->
           let* result =
             Lwt.catch
               (fun () -> handler invocation)
               (fun exn ->
                 Lwt.return
                   (Types.make_tool_result ~result_type:Failure
                      (Printexc.to_string exn)))
           in
           Lwt.return (Types.tool_result_to_yojson result)));
  Jsonrpc.register_handler rpc "permission/handlePermissionRequest"
    (fun _method params ->
      let open Lwt.Syntax in
      match t.permission_handler with
      | None ->
        (* Default: deny *)
        Lwt.return
          (Types.permission_result_to_yojson
             { decision = DeniedByUser; rules = None })
      | Some handler ->
        let* result =
          Lwt.catch
            (fun () -> handler params t.session_id)
            (fun _exn ->
              Lwt.return
                Types.{ decision = DeniedByUser; rules = None })
        in
        Lwt.return (Types.permission_result_to_yojson result));
  Jsonrpc.register_handler rpc "exitPlanMode.request"
    (fun _method params ->
      let open Lwt.Syntax in
      match t.exit_plan_mode_handler with
      | None ->
        Lwt.return
          (Types.exit_plan_mode_response_to_yojson
             (Types.default_exit_plan_mode_response ()))
      | Some handler ->
        (match Types.exit_plan_mode_request_of_yojson params with
         | Error _ ->
           Lwt.return
             (Types.exit_plan_mode_response_to_yojson
                (Types.default_exit_plan_mode_response ()))
         | Ok req ->
           let* result =
             Lwt.catch
               (fun () -> handler req)
               (fun _exn ->
                 Lwt.return (Types.default_exit_plan_mode_response ()))
           in
           Lwt.return (Types.exit_plan_mode_response_to_yojson result)));
  t

(* ========================================================================== *)
(* Event dispatching                                                          *)
(* ========================================================================== *)

let dispatch_event (t : t) (event : Types.session_event) : unit Lwt.t =
  let open Lwt.Syntax in
  (match event.event_type with
   | SessionIdle -> t.idle <- true
   | _ -> t.idle <- false);
  Lwt_list.iter_s (fun handler -> handler event) t.event_handlers

let setup_notifications (t : t) : unit =
  Jsonrpc.set_notification_handler t.rpc
    (fun method_ params ->
      if method_ = "session/event" then
        match Types.session_event_of_yojson params with
        | Ok event -> dispatch_event t event
        | Error _ -> Lwt.return_unit
      else Lwt.return_unit)

(* ========================================================================== *)
(* Public API                                                                 *)
(* ========================================================================== *)

let session_id (t : t) : string = t.session_id

let on (t : t) (handler : event_handler) : unit =
  t.event_handlers <- handler :: t.event_handlers

let inject_trace_context (t : t) (params : Yojson.Safe.t) : Yojson.Safe.t =
  match t.trace_context_provider with
  | None -> params
  | Some provider ->
    (try
       let ctx = provider () in
       let fields = match params with `Assoc fs -> fs | _ -> [] in
       let fields = match ctx.traceparent with
         | Some tp -> ("traceparent", `String tp) :: fields
         | None -> fields
       in
       let fields = match ctx.tracestate with
         | Some ts -> ("tracestate", `String ts) :: fields
         | None -> fields
       in
       `Assoc fields
     with _ -> params)

let send (t : t) (msg : Types.message_options) : unit Lwt.t =
  let params =
    `Assoc
      [ ("sessionId", `String t.session_id)
      ; ("message", Types.message_options_to_yojson msg)
      ]
  in
  let params = inject_trace_context t params in
  let open Lwt.Syntax in
  let* _result = Jsonrpc.send_request t.rpc "session/send" params in
  Lwt.return_unit

let send_and_wait ?(timeout = 60.0) (t : t) (msg : Types.message_options)
    : Types.session_event option Lwt.t =
  let open Lwt.Syntax in
  let waiter, resolver = Lwt.wait () in
  let resolved = ref false in
  let handler (event : Types.session_event) : unit Lwt.t =
    if not !resolved then
      (match event.event_type with
       | AssistantMessage | SessionIdle | SessionError ->
         resolved := true;
         Lwt.wakeup resolver (Some event);
         Lwt.return_unit
       | _ -> Lwt.return_unit)
    else Lwt.return_unit
  in
  on t handler;
  let* () = send t msg in
  let timeout_thread =
    let* () = Lwt_unix.sleep timeout in
    if not !resolved then (
      resolved := true;
      Lwt.wakeup resolver None);
    Lwt.return_unit
  in
  Lwt.async (fun () -> timeout_thread);
  waiter

let register_tool (t : t) (name : string) (handler : tool_handler) : unit =
  Hashtbl.replace t.tool_handlers name handler

let register_permission_handler (t : t) (handler : permission_handler) : unit =
  t.permission_handler <- Some handler

let register_exit_plan_mode_handler (t : t) (handler : exit_plan_mode_handler) : unit =
  t.exit_plan_mode_handler <- Some handler

let register_trace_context_provider (t : t) (provider : unit -> Types.trace_context) : unit =
  t.trace_context_provider <- Some provider

let destroy (t : t) : unit Lwt.t =
  let params = `Assoc [ ("sessionId", `String t.session_id) ] in
  let open Lwt.Syntax in
  let* _result = Jsonrpc.send_request t.rpc "session/destroy" params in
  Lwt.return_unit

let is_idle (t : t) : bool = t.idle

File: fsharp/src/Session.fs
namespace CopilotSDK.Supercharged.FSharp

open System
open System.Collections.Generic
open System.Text.Json
open System.Threading
open System.Threading.Tasks

/// A Copilot conversation session.
///
/// Sessions are created via CopilotClient.createSession or resumed via
/// CopilotClient.resumeSession. They manage event subscriptions, tool
/// registration, and message exchange with the assistant.
type CopilotSession internal (sessionId: string, rpc: JsonRpcTransport, initialWorkspacePath: string option) =

    let eventHandlers = ResizeArray<SessionEvent -> unit>()
    let handlersLock = obj()
    let toolHandlers = Dictionary<string, ToolHandler>()
    let mutable workspacePath = initialWorkspacePath
    let mutable exitPlanModeHandler : ExitPlanModeHandler option = None
    let mutable traceContextProvider : TraceContextProvider option = None

    let dispatchEvent (evt: SessionEvent) =
        let handlers =
            lock handlersLock (fun () -> eventHandlers |> Seq.toList)
        for handler in handlers do
            try handler evt
            with _ -> ()

    do
        // Register for session event notifications from the server.
        rpc.OnNotification("session.event", fun paramData ->
            match paramData with
            | Some elem ->
                let eventType =
                    match elem.TryGetProperty("type") with
                    | true, t -> t.GetString()
                    | _ -> "unknown"
                let data =
                    match elem.TryGetProperty("data") with
                    | true, d -> Some (d.Clone())
                    | _ -> None
                let evt = SessionEvent.parse eventType data
                dispatchEvent evt
            | None -> ()
        )

        // Register for tool call requests from the server.
        rpc.OnNotification("session.toolCall", fun paramData ->
            match paramData with
            | Some elem ->
                let toolName =
                    match elem.TryGetProperty("toolName") with
                    | true, t -> t.GetString()
                    | _ -> ""
                let toolCallId =
                    match elem.TryGetProperty("toolCallId") with
                    | true, t -> t.GetString()
                    | _ -> ""
                let args =
                    match elem.TryGetProperty("arguments") with
                    | true, a -> Some (a.Clone())
                    | _ -> None

                match toolHandlers.TryGetValue(toolName) with
                | true, handler ->
                    let invocation =
                        { SessionId = sessionId
                          ToolCallId = toolCallId
                          ToolName = toolName
                          Arguments = args }
                    async {
                        try
                            let! result = handler invocation
                            let resultObj =
                                match result with
                                | :? ToolResultObject as tro -> tro
                                | :? string as s -> ToolResultObject.success s
                                | other ->
                                    let json = JsonSerializer.Serialize(other)
                                    ToolResultObject.success json
                            do! rpc.SendNotificationAsync(
                                "session.toolCallResult",
                                {| sessionId = sessionId
                                   toolCallId = toolCallId
                                   result = resultObj |})
                        with ex ->
                            do! rpc.SendNotificationAsync(
                                "session.toolCallResult",
                                {| sessionId = sessionId
                                   toolCallId = toolCallId
                                   result = ToolResultObject.failure ex.Message |})
                    } |> Async.Start
                | _ -> ()
            | None -> ()
        )

        // Register handler for exit plan mode requests from the server.
        rpc.OnRequest("exitPlanMode.request", fun paramData ->
            async {
                match paramData with
                | Some elem ->
                    let summary =
                        match elem.TryGetProperty("summary") with
                        | true, s -> s.GetString()
                        | _ -> ""
                    let planContent =
                        match elem.TryGetProperty("planContent") with
                        | true, s -> Some (s.GetString())
                        | _ -> None
                    let actions =
                        match elem.TryGetProperty("actions") with
                        | true, a ->
                            [ for i in 0 .. a.GetArrayLength() - 1 -> a.[i].GetString() ]
                        | _ -> []
                    let recommendedAction =
                        match elem.TryGetProperty("recommendedAction") with
                        | true, s -> s.GetString()
                        | _ -> ""
                    let request =
                        { SessionId = sessionId
                          Summary = summary
                          PlanContent = planContent
                          Actions = actions
                          RecommendedAction = recommendedAction }
                    let! result =
                        match exitPlanModeHandler with
                        | None ->
                            async { return { Approved = true; SelectedAction = None; Feedback = None } }
                        | Some handler ->
                            async {
                                try
                                    return! handler request
                                with _ ->
                                    return { Approved = true; SelectedAction = None; Feedback = None }
                            }
                    return
                        {| approved = result.Approved
                           selectedAction = result.SelectedAction |> Option.toObj
                           feedback = result.Feedback |> Option.toObj |} :> obj
                | None ->
                    return {| approved = true |} :> obj
            })

    /// Alternate constructor without workspace path.
    internal new(sessionId: string, rpc: JsonRpcTransport) =
        CopilotSession(sessionId, rpc, None)

    /// The unique identifier for this session.
    member _.SessionId = sessionId

    /// The workspace path if infinite sessions are enabled.
    member _.WorkspacePath = workspacePath

    /// Subscribe to all session events. Returns an IDisposable that
    /// removes the handler when disposed.
    ///
    /// Use pattern matching in F# to handle events idiomatically:
    ///   session.On(function
    ///       | AssistantMessage data -> printfn "%s" data.Content
    ///       | SessionIdle -> printfn "Done"
    ///       | _ -> ())
    member _.On(handler: SessionEvent -> unit) : IDisposable =
        lock handlersLock (fun () -> eventHandlers.Add(handler))
        { new IDisposable with
            member _.Dispose() =
                lock handlersLock (fun () -> eventHandlers.Remove(handler) |> ignore) }

    /// Subscribe to session events via an IObservable.
    member this.Events : IObservable<SessionEvent> =
        { new IObservable<SessionEvent> with
            member _.Subscribe(observer) =
                this.On(fun evt -> observer.OnNext(evt)) }

    /// Register a tool that the assistant can invoke during this session.
    member _.RegisterTool(registeredTool: RegisteredTool) =
        toolHandlers.[registeredTool.Tool.Name] <- registeredTool.Handler

    /// Register multiple tools at once.
    member this.RegisterTools(tools: RegisteredTool list) =
        tools |> List.iter this.RegisterTool

    /// Register a handler for exit plan mode requests.
    member _.RegisterExitPlanModeHandler(handler: ExitPlanModeHandler) =
        exitPlanModeHandler <- Some handler

    /// Register a trace context provider for distributed tracing.
    member _.RegisterTraceContextProvider(provider: TraceContextProvider) =
        traceContextProvider <- Some provider

    /// Send a message to the session. Returns the response message ID.
    member _.SendAsync(options: MessageOptions, ?cancellationToken: CancellationToken) : Async<string> = async {
        let request = Dictionary<string, obj>()
        request.["sessionId"] <- box sessionId
        request.["prompt"] <- box options.Prompt
        request.["attachments"] <- box (options.Attachments |> Option.defaultValue [])
        options.Mode |> Option.iter (fun m -> request.["mode"] <- box m)
        match traceContextProvider with
        | Some provider ->
            try
                let ctx = provider ()
                ctx.Traceparent |> Option.iter (fun tp -> request.["traceparent"] <- box tp)
                ctx.Tracestate |> Option.iter (fun ts -> request.["tracestate"] <- box ts)
            with _ -> ()
        | None -> ()
        let! response = rpc.SendRequestAsync<{| messageId: string |}>("session.send", request)
        return response.messageId
    }

    /// Send a message and wait until the session becomes idle.
    /// Returns the final assistant message, or None if no message was received.
    member this.SendAndWaitAsync(
            options: MessageOptions,
            ?timeout: TimeSpan,
            ?cancellationToken: CancellationToken) : Async<AssistantMessageData option> = async {
        let effectiveTimeout = defaultArg timeout (TimeSpan.FromSeconds(60.0))
        let tcs = TaskCompletionSource<AssistantMessageData option>()
        let mutable lastMessage : AssistantMessageData option = None

        use _ = this.On(fun evt ->
            match evt with
            | AssistantMessage data ->
                lastMessage <- Some data
            | SessionIdle ->
                tcs.TrySetResult(lastMessage) |> ignore
            | _ -> ())

        let! _messageId = this.SendAsync(options)

        use ctSource = new CancellationTokenSource(effectiveTimeout)
        use _reg = ctSource.Token.Register(fun () ->
            tcs.TrySetException(TimeoutException("SendAndWait timed out")) |> ignore)

        return! tcs.Task |> Async.AwaitTask
    }

    /// Retrieve the conversation history for this session.
    member _.GetHistoryAsync(?cancellationToken: CancellationToken) : Async<JsonElement> = async {
        return! rpc.SendRequestAsync<JsonElement>("session.getHistory", {| sessionId = sessionId |})
    }

    /// Stop the session (notifies the server).
    member _.StopAsync(?cancellationToken: CancellationToken) : Async<unit> = async {
        do! rpc.SendNotificationAsync("session.stop", {| sessionId = sessionId |})
    }
